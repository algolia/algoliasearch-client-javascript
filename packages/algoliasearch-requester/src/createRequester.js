// @flow

import RequestHosts from './hosts';

import type { AppId, ApiKey } from 'types/Algolia';
import type {
  RequestOptions,
  RequestArguments,
  HttpModule,
  Timeouts,
  Hosts,
} from 'algoliasearch-requester';

type Args = {|
  appId?: AppId,
  apiKey: ApiKey,
  httpRequester: HttpModule,
  options?: {|
    timeouts?: Timeouts,
    extraHosts?: Hosts,
  |},
  requestOptions?: RequestOptions,
|};

const stringify = qs => JSON.stringify(qs); // todo: use proper url stringify

type ErrorTypes = 'application' | 'network' | 'dns' | 'timeout';
const retryableErrors: Array<ErrorTypes> = [
  'application',
  'network',
  'dns',
  'timeout',
];

const RESET_HOST_TIMER = 12000; // ms; 2 minutes
const RESET_TIMEOUT_TIMER = 120000; // ms; 20 minutes

export class Requester {
  hosts: RequestHosts;
  apiKey: ApiKey;
  appId: AppId;
  requestOptions: RequestOptions;
  requester: HttpModule;

  constructor({
    appId,
    apiKey,
    httpRequester,
    options = {},
    requestOptions = {},
  }: Args) {
    if (typeof appId !== 'string') {
      throw new Error(
        `appId is required and should be a string, received ${appId}`
      );
    }
    if (typeof apiKey !== 'string') {
      throw new Error(
        `apiKey is required and should be a string, received ${apiKey}`
      );
    }
    if (typeof httpRequester !== 'function') {
      throw new Error(
        `httpRequester is required and should be a function, received ${httpRequester}`
      );
    }
    this.hosts = new RequestHosts({ appId, ...options });
    this.appId = appId;
    this.apiKey = apiKey;
    this.requester = httpRequester;
    this.requestOptions = requestOptions;
  }

  setOptions = (fn: RequestOptions => RequestOptions): RequestOptions => {
    const oldOptions = this.requestOptions;
    const newOptions = fn(oldOptions);
    this.requestOptions = newOptions;
    return newOptions;
  };

  request = ({
    method,
    path,
    qs,
    body,
    options,
    requestType: type,
    retry = 0,
  }: RequestArguments) => {
    const hostname = this.hosts.getHost({ type });
    const timeout = this.hosts.getTimeout({ retry, type });

    const pathname = path + stringify(qs);
    const url = { hostname, pathname };

    return this.requester({
      body,
      method,
      url,
      timeout,
      options,
    }).catch(err =>
      this.retryRequest({
        err,
        method,
        path,
        qs,
        body,
        options,
        type,
        retry,
      })
    );
  };

  retryRequest = ({ err, method, path, qs, body, options, type, retry }) => {
    console.warn(err.message);
    if (retryableErrors.indexOf(err.message.reason) > -1) {
      // if no more hosts or timeouts: reject
      // if reason: timeout; increase
      return this.request({
        method,
        path,
        qs,
        body,
        options,
        requestType: type,
        retry: retry + 1,
      });
    }
    return Promise.reject(
      new Error({
        reason:
          "Request couldn't be retried, did you enter the correct credentials?",
        more: err.message.reason,
      })
    );
  };
}

export default function createRequester(args: Args) {
  const _r = new Requester(args);
  const requester = _r.request;
  requester.setOptions = _r.setOptions;
  requester.options = _r.requestOptions;
  return requester;
}
