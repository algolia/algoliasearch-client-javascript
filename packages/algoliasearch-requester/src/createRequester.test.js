import createRequester from './createRequester';

jest.useFakeTimers();

it('requires the right arguments', () => {
  const falseInvocations = [
    () =>
      createRequester({
        apiKey: '',
      }),
    () =>
      createRequester({
        appId: '',
      }),
    () =>
      createRequester({
        appId: '',
        apiKey: '',
        httpRequester: {},
      }),
  ];

  falseInvocations.map(invocation =>
    expect(invocation).toThrowErrorMatchingSnapshot()
  );

  expect(() =>
    createRequester({
      appId: '',
      apiKey: '',
      httpRequester: () => {},
    })
  ).not.toThrow();
});

it('first read request uses first host', () => {
  const httpRequester = jest.fn(() => Promise.resolve());
  const requester = createRequester({
    appId: 'the_read_app',
    apiKey: '',
    httpRequester,
  });

  requester({
    requestType: 'read',
  });

  const firstArgs = httpRequester.mock.calls[0];
  const { url: { hostname } } = firstArgs[0];
  expect(hostname).toEqual('the_read_app-dsn.algolia.net');
});

it('first write request uses first host', () => {
  const httpRequester = jest.fn(() => Promise.resolve());
  const requester = createRequester({
    appId: 'the_write_app',
    apiKey: '',
    httpRequester,
  });

  requester({
    requestType: 'write',
  });

  const firstArgs = httpRequester.mock.calls[0];
  const { url: { hostname } } = firstArgs[0];
  expect(hostname).toEqual('the_write_app.algolia.net');
});

it.skip('uses a different host when the request needs to be retried', () => {
  const httpRequester = jest.fn(
    () =>
      httpRequester.mock.calls.length === 1
        ? Promise.reject(new Error({ reason: 'network' }))
        : Promise.resolve()
  );
  const requester = createRequester({
    appId: 'the_crazy_app',
    apiKey: '',
    httpRequester,
  });

  requester({
    requestType: 'read',
  }); // retries

  const usedHosts = [
    httpRequester.mock.calls[0][0].url.hostname,
    httpRequester.mock.calls[1][0].url.hostname,
    httpRequester.mock.calls[2][0].url.hostname,
  ];
  expect(usedHosts[0]).toEqual('the_crazy_app-dsn.algolia.net'); // first try
  expect(usedHosts[1]).toEqual('the_crazy_app-1.algolianet.com'); // second try
});

it.skip('uses the "up" host on second request when first fails', () => {
  const httpRequester = jest.fn(
    () =>
      httpRequester.mock.calls.length === 1
        ? Promise.reject(new Error({ reason: 'network' }))
        : Promise.resolve()
  );
  const requester = createRequester({
    appId: 'the_crazy_app',
    apiKey: '',
    httpRequester,
  });

  requester({
    requestType: 'read',
  }); // retries
  requester({
    requestType: 'read',
  });

  const usedHosts = [
    httpRequester.mock.calls[0][0].url.hostname,
    httpRequester.mock.calls[1][0].url.hostname,
    httpRequester.mock.calls[2][0].url.hostname,
  ];
  expect(usedHosts[0]).toEqual('the_crazy_app-dsn.algolia.net'); // first try
  expect(usedHosts[1]).toEqual('the_crazy_app-1.algolianet.com'); // first retry
  expect(usedHosts[2]).toEqual('the_crazy_app-1.algolianet.com'); // second request
});

it('resolves when the response is successful', () => {
  const httpRequester = jest.fn(() => Promise.resolve({}));
  const requester = createRequester({
    appId: 'the_successful_app',
    apiKey: '',
    httpRequester,
  });

  expect(
    requester({
      requestType: 'write',
    })
  ).resolves.toEqual({});
});

it.skip("retries when there's an application error", () => {
  const httpRequester = jest.fn(
    () =>
      httpRequester.mock.calls.length === 1
        ? Promise.reject(
            new Error({
              reason: 'application',
            })
          )
        : Promise.resolve({})
  );
  const requester = createRequester({
    appId: 'the_app_app',
    apiKey: '',
    httpRequester,
  });

  // it eventually resolves
  expect(
    requester({
      requestType: 'write',
    })
  ).resolves.toEqual({});

  // requester was called twice
  expect(httpRequester.mock.calls).toHaveLength(2);
});

it.skip("retries when there's a network error", () => {
  const httpRequester = jest.fn(
    () =>
      httpRequester.mock.calls.length === 1
        ? Promise.reject(
            new Error({
              reason: 'network',
            })
          )
        : Promise.resolve({})
  );
  const requester = createRequester({
    appId: 'the_network_app',
    apiKey: '',
    httpRequester,
  });

  // it eventually resolves
  expect(
    requester({
      requestType: 'write',
    })
  ).resolves.toEqual({});

  // requester was called twice
  expect(httpRequester.mock.calls).toHaveLength(2);
});

it.skip("retries when there's a dns error", () => {
  const httpRequester = jest.fn(
    () =>
      httpRequester.mock.calls.length === 1
        ? Promise.reject(
            new Error({
              reason: 'dns',
            })
          )
        : Promise.resolve({})
  );
  const requester = createRequester({
    appId: 'the_dns_app',
    apiKey: '',
    httpRequester,
  });

  // it eventually resolves
  expect(
    requester({
      requestType: 'write',
    })
  ).resolves.toEqual({});

  // requester was called twice
  expect(httpRequester.mock.calls.length).toBe(2);
});

it.skip("retries when there's a timeout", () => {
  const httpRequester = jest.fn(
    () =>
      httpRequester.mock.calls.length === 1
        ? Promise.reject(
            new Error({
              reason: 'timeout',
            })
          )
        : Promise.resolve({})
  );
  const requester = createRequester({
    appId: 'the_retry_app',
    apiKey: '',
    httpRequester,
  });

  // it eventually resolves
  expect(
    requester({
      requestType: 'write',
    })
  ).resolves.toEqual({});

  // requester was called twice
  expect(httpRequester.mock.calls.length).toBe(2);
});

it.skip('second try after a timeout has increments the timeout', () => {
  const httpRequester = jest.fn(
    () =>
      httpRequester.mock.calls.length === 1
        ? Promise.reject(
            new Error({
              reason: 'timeout',
            })
          )
        : Promise.resolve({})
  );
  const requester = createRequester({
    appId: 'the_fun_app',
    apiKey: '',
    httpRequester,
  });

  requester({
    requestType: 'write',
  });

  expect(/* the current timeout */).toBeGreaterThan(/* the original timeout */);
});

it.skip('rejects when all timeouts are reached', () => {
  const httpRequester = jest.fn(() =>
    Promise.reject(
      new Error({
        reason: 'timeout',
      })
    )
  );
  const requester = createRequester({
    appId: 'the_timeout_app',
    apiKey: '',
    httpRequester,
  });

  expect(() => {
    requester({
      requestType: 'write',
    }); // eventually it will fail because of a timeout
  }).rejects.toBe(new Error('replace with the total timeout error'));

  requester({ requestType: 'write' });

  const calls = httpRequester.mock.calls;
  const firstTimeout = calls[0][0].timeout;
  const lastTimeout = calls[calls.length - 1][0].timeout;

  expect(lastTimeout).toBe(firstTimeout);
});

it.skip('rejects when all hosts are used', () => {
  const httpRequester = jest.fn(() =>
    Promise.reject(
      new Error({
        reason: 'host',
      })
    )
  );
  const requester = createRequester({
    appId: 'the_host_app',
    apiKey: '',
    httpRequester,
  });

  // eventually it will fail because there are no more hosts
  expect(() =>
    requester({
      requestType: 'write',
    })
  ).rejects.toBe(new Error('replace with the host error'));
});

it.skip('uses the first host again after throwing', () => {
  const httpRequester = jest.fn(
    () =>
      httpRequester.mock.calls.length < 4 /* the request completely fails */
        ? Promise.reject(
            new Error({
              reason: 'network',
            })
          )
        : Promise.resolve({})
  );
  const requester = createRequester({
    appId: 'the_other_host_app',
    apiKey: '',
    httpRequester,
  });

  // eventually it will fail because there are no more hosts
  expect(() =>
    requester({
      requestType: 'write',
    })
  ).rejects.toBe(new Error('replace with the host error'));

  requester({
    requestType: 'write',
  }); // this request works

  const calls = httpRequester.mock.calls;
  const firstHost = calls[0][0].hostname;
  const lastHost = calls[calls.length - 1][0].hostname;

  expect(lastHost).toBe(firstHost);
});

it.skip('two instances of createRequester share the same host index', () => {
  const httpRequester = jest.fn(
    () =>
      httpRequester.mock.calls.length === 1
        ? Promise.reject(new Error({ reason: 'network' }))
        : Promise.resolve()
  );

  const firstRequester = createRequester({
    appId: 'the_same_app',
    apiKey: '',
    httpRequester,
  });
  const secondRequester = createRequester({
    appId: 'the_same_app',
    apiKey: '',
    httpRequester,
  });

  firstRequester({
    requestType: 'read',
  }); // retries
  secondRequester({
    requestType: 'read',
  });

  const usedHosts = [
    httpRequester.mock.calls[0][0].url.hostname,
    httpRequester.mock.calls[1][0].url.hostname,
    httpRequester.mock.calls[2][0].url.hostname,
  ];
  expect(usedHosts[0]).toEqual('the_same_app-dsn.algolia.net'); // first try
  expect(usedHosts[1]).toEqual('the_same_app-1.algolianet.com'); // first retry
  expect(usedHosts[2]).toEqual('the_same_app-1.algolianet.com'); // second request
});

it.skip('host indices are reset to 0 after Xs', () => {
  const httpRequester = jest.fn(
    () =>
      httpRequester.mock.calls.length === 1
        ? Promise.reject(new Error({ reason: 'network' }))
        : Promise.resolve()
  );
  const requester = createRequester({
    appId: 'the_slow_app',
    apiKey: '',
    httpRequester,
  });

  requester({
    requestType: 'read',
  }); // retries

  // wait X seconds
  requester({
    requestType: 'read',
  }); // retries

  const usedHosts = [
    httpRequester.mock.calls[0][0].url.hostname,
    httpRequester.mock.calls[2][0].url.hostname,
    httpRequester.mock.calls[3][0].url.hostname,
  ];
  expect(usedHosts[0]).toEqual('the_slow_app-dsn.algolia.net'); // first try
  expect(usedHosts[1]).toEqual('the_slow_app-1.algolianet.com'); // second try
  expect(usedHosts[2]).toEqual('the_slow_app-dsn.algolia.net'); // third try
});
