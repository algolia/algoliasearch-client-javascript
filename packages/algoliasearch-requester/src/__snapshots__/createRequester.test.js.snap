// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`requires the right arguments 1`] = `"appId is required and should be a string, received undefined"`;

exports[`requires the right arguments 2`] = `"apiKey is required and should be a string, received undefined"`;

exports[`requires the right arguments 3`] = `"httpRequester is required and should be a function, received [object Object]"`;
